File: apps/web/src/server/routers/util/proto/campaign.ts
import { timestampDate } from "@bufbuild/protobuf/wkt";
import { ORPCError } from "@orpc/client";
import { type Campaign, CampaignSchema } from "@planner/schemas/campaigns";
import type { Campaign as CampaignProto } from "@/gen/proto/planner/v1/campaign_pb";

export function protoToCampaign(proto: CampaignProto): Campaign {
	if (!proto.createdAt) {
		throw new ORPCError("INTERNAL_SERVER_ERROR", {
			message: "CampaignInvitation is missing createdAt",
		});
	}
	if (!proto.updatedAt) {
		throw new ORPCError("INTERNAL_SERVER_ERROR", {
			message: "CampaignInvitation is missing updatedAt",
		});
	}

	return CampaignSchema.parse({
		createdAt: timestampDate(proto.createdAt),
		deletedAt: proto.deletedAt ? timestampDate(proto.deletedAt) : null,
		description: proto.description,
		id: proto.id,
		tags: proto.tags,
		title: proto.title,
		updatedAt: timestampDate(proto.updatedAt),
		userId: proto.userId,
	});
}

File: apps/api/internal/models/campaign.go
package model

import (
	"database/sql"
	"time"
)

type Campaign struct {
	ID          string
	UserID      string
	Title       string
	Description sql.NullString
	Tags        []string
	CreatedAt   time.Time
	UpdatedAt   time.Time
	DeletedAt   sql.NullTime
}

type CreateCampaignRequest struct {
	UserID      string
	Title       string
	Description sql.NullString
	Tags        []string
}

type UpdateCampaignRequest struct {
	ID          string
	Title       *string
	Description sql.NullString
	Tags        []string
}

type CampaignAuth struct {
	*Campaign
	ColonyID *string
}
